app/lib/expression_viz_service.rb

```diff
@@ -107,121 +107,35 @@ def self.load_expression_boxplot_data_array_scores(study, gene, cluster, annotat
   end
 
   # method to load a 2-d scatter of selected numeric annotation vs. gene expression
-  def self.load_annotation_based_data_array_scatter(study, gene, cluster, annotation, subsample_threshold, y_axis_title)
+  def self.load_annotation_based_data_array_scatter(study, gene, cluster, annotation, subsample_threshold)
 
     # construct annotation key to load subsample data_arrays if needed, will be identical to params[:annotation]
     subsample_annotation = "#{annotation[:name]}--#{annotation[:type]}--#{annotation[:scope]}"
     cells = cluster.concatenate_data_arrays('text', 'cells', subsample_threshold, subsample_annotation)
-    annotation_array = []
-    annotation_hash = {}
-    if annotation[:scope] == 'cluster'
-      annotation_array = cluster.concatenate_data_arrays(annotation[:name], 'annotations', subsample_threshold, subsample_annotation)
-    elsif annotation[:scope] == 'user'
-      # for user annotations, we have to load by id as names may not be unique to clusters
-      user_annotation = UserAnnotation.find(annotation[:id])
-      subsample_annotation = user_annotation.formatted_annotation_identifier
-      annotation_array = user_annotation.concatenate_user_data_arrays(annotation[:name], 'annotations', subsample_threshold, subsample_annotation)
-      cells = user_annotation.concatenate_user_data_arrays('text', 'cells', subsample_threshold, subsample_annotation)
-    else
-      # for study-wide annotations, load from study_metadata values instead of cluster-specific annotations
-      metadata_obj = study.cell_metadata.by_name_and_type(annotation[:name], annotation[:type])
-      annotation_hash = metadata_obj.cell_annotations
-    end
-    annotation_array = AnnotationVizService.sanitize_values_array(annotation_array, annotation[:type])
-    values = {}
-    values[:all] = {x: [], y: [], cells: [], annotations: [], text: [], marker: {size: study.default_cluster_point_size,
-      line: { color: 'rgb(40,40,40)', width: study.show_cluster_point_borders? ? 0.5 : 0}}}
-    if annotation[:scope] == 'cluster' || annotation[:scope] == 'user'
-      annotation_array.each_with_index do |annot, index|
-        annotation_value = annot
-        cell_name = cells[index]
-        expression_value = gene['scores'][cell_name].to_f.round(4)
+    annotation_array = ClusterVizService.get_annotation_values_array(study, cluster, annotation, cells, subsample_annotation, subsample_threshold)
```

Comment:
Same concern as above with not sanitizing this array.

Reply:
The sanitization happens inside that utility function